### ISSUE TITLE
Register custom UrlReporter for Cucumber Reports

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/cucumber-jvm#2324]
**Is your feature request related to a problem? Please describe.**

I would like to capture the URL of the published report and send a message to an external service (Slack) with the URL of the report. 

**Describe the solution you'd like**

I want to be able to specify an alternative implementation of `UrlReporter`, perhaps using a system property:

```
-Dcucumber.publish.urlreporter=com.acme.MyUrlReporter 
```

Work involved:

* Split the current `UrlReporter` class to a `UrlReporter` interface and a `DefaultUrlReporter` class
* Use `java.util.ServiceLoader` to instantiate the class, taken from the `cucumber.publish.urlreporter` system property

**Describe alternatives you've considered**

Redirecting STDERR to a file, then parse it with a shell script and send the message. That's quite a roundabout way of doing this.

**Additional context**

This is actually something that's requested by a user of Cucumber Reports.

**Comment by mpkorstanje** [comment]
I imagine you may be better of by making Cucumber reports a Slack app that makes the call backs. Similar to the Jira and Pagerduty Slack apps.

**Comment by mpkorstanje** [comment]
Okay. I am definitely convinced this isn't a good feature. Aside from being a nice example of feature creep, if the cucumber reports app were to make the calls to slack it would allow this functionality to work for all implementations rather then only `cucumber-jvm`.

**Comment by aslakhellesoy** [comment]
I wasn't suggesting implementing the slack logic - I only mentioned slack as an example. All I'm suggesting is to allow users to specify an implementation class of their own.

It could be used to send the result to a variety of destinations, depending on the users' needs. Some possible use cases:

* Send a request to a HTTP API
* Append to a file
* Send an email

The idea is to open up the API so users can capture the URL in their own team/environment/organisation specific way.

**Comment by mpkorstanje** [comment]
I'm aware. I don't think that is worth the complexity. Ideally we provide a plugin system such that these concerns are addressable by the end user. And between a redesign, feature creep or alternative solution I reckon the alternative solution is the better way to support the example usecase.

**Comment by mpkorstanje** [comment]
See also: https://github.com/cucumber/cucumber-jvm/issues/1834

**Comment by aslakhellesoy** [comment]
I agree we should have a uniform plugin solution rather than many different ones sprinkled across the codebase.

Are you saying that if we implement #1834 we might be able to support a pluggable `UrlReporter` too?

**Comment by mpkorstanje** [comment]
No but the publish plugin would have it's own module and it's own api. Part of that API could be a client to talk to the reports api.

People with different use cases could reuse that client in their own plugin. And then for example compose their slack notifier around it.

The biggest challenge is accessing configuration for plugins and url reporters. With a better plugin system we can make that easier. Instead of providing a single purpose extension point  by enabling composition we can facilitate much more.

**Comment by ahmetchakmak** [comment]
I am very surprised this feature is not implemented already. Clicking on a console output shouldn't be the only way to reach report. There is basically no way to share report link with others with your automation code.

**Comment by aslakhellesoy** [comment]
Also see the last paragraph of the [LICENSE](https://github.com/cucumber/cucumber-jvm/blob/main/LICENCE). If you really want this you can contribute it yourself. We're here to help you if you have any questions about how to proceed.
